[HttpKernel] Add `#[MapQueryParameter]`, `#[MapQueryString]` and `#[MapRequestPayload]` (#18336)
symfony/symfony-docs: controller.rst

    // ...

    public function dashboard(
        #[MapQueryParameter] string $firstName,
        #[MapQueryParameter] string $lastName,
        #[MapQueryParameter] int $age,
    ): Response
    {
        // ...
    }

``#[MapQueryParameter]`` can take an optional argument called ``filter``. You can use the
`Validate Filters`_ constants defined in PHP::

    use Symfony\Component\HttpFoundation\Response;
    use Symfony\Component\HttpKernel\Attribute\MapQueryParameter;

    // ...

    public function dashboard(
        #[MapQueryParameter(filter: \FILTER_VALIDATE_REGEXP, options: ['regexp' => '/^\w++$/'])] string $firstName,
        #[MapQueryParameter] string $lastName,
        #[MapQueryParameter(filter: \FILTER_VALIDATE_INT)] int $age,
    ): Response
    {
        // ...
    }

.. versionadded:: 6.3

    The :class:`Symfony\\Component\\HttpKernel\\Attribute\\MapQueryParameter` attribute
    was introduced in Symfony 6.3.

Mapping The Whole Query String
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Another possibility is to map the entire query string into an object that will hold
available query parameters. Let's say you declare the following DTO with its

> validation constraints::

Reviewer: ```suggestion
optional validation constraints::
```

to decouple the fact that we can map without any constraint rules :)

Author: Good suggestion in my opinion, added! :slightly_smiling_face:

    class UserDTO
    {
        public function __construct(
            #[Assert\NotBlank]
            public string $firstName,

            #[Assert\NotBlank]
            public string $lastName,

            #[Assert\GreaterThan(18)]
            public int $age,
        ) {
        }
    }

You can then use the :class:`Symfony\\Component\\HttpKernel\\Attribute\\MapQueryString`
attribute in your controller::

    use App\Model\UserDto;
    use Symfony\Component\HttpFoundation\Response;
    use Symfony\Component\HttpKernel\Attribute\MapQueryString;

    // ...

    public function dashboard(
        #[MapQueryString] UserDTO $userDto
    ): Response
    {
        // ...
    }

.. versionadded:: 6.3

    The :class:`Symfony\\Component\\HttpKernel\\Attribute\\MapQueryString` attribute
    was introduced in Symfony 6.3.

Mapping Request Payload
~~~~~~~~~~~~~~~~~~~~~~~

When creating an API and dealing with other HTTP methods than ``GET`` (like
``POST`` or ``PUT``), user's data are not stored in the query string
but directly in the request payload, like this:

.. code-block:: json

    {
        "firstName": "John",
        "lastName": "Smith",
        "age": 28
    }

> In this case, it is also possible to directly map this payload to your DTO thanks
> to the :class:`Symfony\\Component\\HttpKernel\\Attribute\\MapRequestPayload`

Reviewer 2: ```suggestion
In this case, it is also possible to directly map this payload to your DTO by
using the :class:`Symfony\\Component\\HttpKernel\\Attribute\\MapRequestPayload`
```

attribute::

    use App\Model\UserDto;
    use Symfony\Component\HttpFoundation\Response;
    use Symfony\Component\HttpKernel\Attribute\MapRequestPayload;

    // ...

    public function dashboard(
        #[MapRequestPayload] UserDTO $userDto
    ): Response
    {
        // ...
    }

This attribute allows you to customize the serialization context as well

> as the class responsible of doing the mapping between the request and

Reviewer 2: ```suggestion
as the class is responsible of doing the mapping between the request and
```

Author: I'm not sure about this :thinking: The sentence seems correct as is. Or maybe `as the class that is responsible`?

Reviewer 3: I agree with Alex on this.
